### PR TITLE
Fix #819: Release process should automate the 'release' stage of the release process down to a single manual step

### LINKED ISSUE #819: Release process should automate the 'release' stage of the release process down to a single manual step
**Is your feature request related to a problem? Please describe.**

We currently have a `stage` workflow step which involves running a workflow action followed by a `release`.

1. The `stage` step branch/tags the code and stages the release artefacts.  
2. The `release` step releases the maven artefacts and merges a release PR.  This involves to developer interactions - one to kick off the workflow, one to merge the PR.   Merging the PR needs to be done in the ["right" way](https://github.com/kroxylicious/kroxylicious/blob/main/RELEASING.md#making-the-release-public) too.

It'd be nice if 2. could be reduced to a single manual interaction.

**Describe the solution you'd like**

How about if the step 2 was invoked by adding a comment to the PR. Something like:

```
/releaser release
```

or 
```
/releaser drop
```

That'd trigger the release workflow with the right argument. In addition would merge (or close) the release PR.

In addition, the stage workflow step could comment on the PR telling the developer what they need to add to the POM to test the stage artefacts.  I mean add a comment that look like this:

_Hey, that's your release successfully staged to Nexus.  It is not public available on Maven Central yet.  If you'd like to test the staged artefacts from a Maven project, you can do that by adding a `<repository>` section to your pom like the below and updating your  `io.kroxylicious:kroxylicious-bom` dependency to version `0.10.0`._

```

   <repositories>
        <repository>
            <releases>
                <enabled>true</enabled>
                <updatePolicy>always</updatePolicy>
                <checksumPolicy>warn</checksumPolicy>
            </releases>
            <name>Nexus Stage</name>
            <id>snapshots-repo</id>
            <url>https://s01.oss.sonatype.org/content/repositories/iokroxylicious-1047</url>
            <layout>default</layout>
        </repository>
    </repositories>

```

_Once you've done testing come back here and add `/releaser release` to release the artefacts, or `/release drop` to if need to abandon it._

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Step towards #819.  

Second stage of the release is now performed by commenting on the release PR.  This will release the staged artifacts to Maven Central and then merge the release PR in the correct manner.      If you want to drop the release there's the robot understand a drop command too.

`@kroxylicious-robot (promote-release|drop-release)`

There are a couple of checks built in to stop anyone using the robot.  Only **Developers** may interact with the robot.  If you are not in the Developer team you'll get an error message posted back.  Only **release-engineers** may actually promote and drop the release.  

Promote has a couple of other safety checks:

* that the branch actually looks like a release branch.
* that PR and main haven't diverged.
* the PR workflow for the the maven run actually finished successfully

I've also used a https://docs.github.com/en/rest/commits/statuses?apiVersion=2022-11-28 to mean that merge button will be disabled on the PR until to the artefacts are released.

### Additional Context

I used lots of features of Github Actions that I haven't used before, so there could well be a better way to do it.

I've tested the drop path pretty well on my fork.  There's some paths that I can't test until it gets merged to main (the permissions check and the actual merge of the PR itself.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [inline .github/workflows/robot-command-dispatcher.yaml:28]: I haven't been able to test the team membership check.  This will need to be tested once it gets merged to main.

File: .github/workflows/deploy_release.yaml
Comment (k-wall) [inline .github/workflows/deploy_release.yaml:105]: I want to fail here if main has move.  I am thinking of using `--match-head-commit` with a SHA calculated earlier in the workflow.

File: .github/workflows/deploy_release.yaml
Comment (k-wall) [inline .github/workflows/deploy_release.yaml:70]: I want a check here that the head of main hasn't moved since the release was staged.  I haven't figured out the exactly `git` incantation.

Comment (SamBarker) [reply .github/workflows/deploy_release.yaml:70]: I've got no idea what the correct git interaction would be but I know we can enable checking that the branch is fully up to date with main as part of the branch protection rules.

File: .github/workflows/deploy_release.yaml
Comment (SamBarker) [inline .github/workflows/deploy_release.yaml:108]: Why the compound condition? `always()` seems like an odd thing to see in the if. I presume this is just a testing hangover.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:34]: I'm fine with `admins` for now, but we should create a release team or grant all members of the `engineering` team release rights.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:44]: When I saw `robot-command-dispatcher` I was hoping for some better pre-built options. Happy this does what we need but the support seems a bit lacking.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:49]: I realise that these are the commands we have as things stand but given the generic nature of `kroxylicious-robot` I think we should start of with a more explicit set of commands. 

I'm thinking some thing along the lines of
- `drop-release` or `drop release` 
- `promote-release` or  `promote-release` 

`release` just feels ambiguous and arguably the release has already happened (as thats the thing that got staged)  and at this point we are promoting the staged release to be public.    

I'm happy to leave `drop` and `release` as the args to the scripts as they are implantation details rather than public API (so to speak)

**Review by SamBarker — COMMENTED** [review]
I was curious as to what all the email spam over the weekend was about. 

Thanks for picking this up @k-wall I think it will be a very helpful starting point for automation.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:34]: Yes, agree.   I had the same idea.   I wanted to get some feedback on the rough direction of the PR before adding the team.  I'll do it.

File: .github/workflows/deploy_release.yaml
Comment (k-wall) [reply .github/workflows/deploy_release.yaml:70]: That sounds like a good idea.  I think I still want to preflight the promote.  We want to know that the PR will apply before we start the promoting the maven artefacts.

I have a simple idea for the preflight check.

File: .github/workflows/deploy_release.yaml
Comment (k-wall) [reply .github/workflows/deploy_release.yaml:108]: My intent was to drop the PR even if, say the deletion of the release notes fails.   I can express this better by make the transition release notes drop path idempotent.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:44]: I was trying to act of your feedback you left on the issue.  You expressed a wish for something generic.   In the PR, I'm trying to keep the dispatch workflow command neutral and leave it up to the sub flows to do the work.

Comment (SamBarker) [reply .github/workflows/robot-command-dispatcher.yaml:44]: Sorry I meant I was expecting something better in terms of bot support from GitHub not this PR! 

I think what you've got is absolutely fine.

File: .github/workflows/stage_release.yaml
Comment (k-wall) [inline .github/workflows/stage_release.yaml:153]: We said we want to automate this part, but let's leave that out of scope from this PR.

File: .github/workflows/stage_release.yaml
Comment (k-wall) [inline .github/workflows/stage_release.yaml:156]: I've used HTML entities here to stop the robot responding to this comment.

File: .github/workflows/deploy_release.yaml
Comment (k-wall) [reply .github/workflows/deploy_release.yaml]: The transition release notes drop path is now idempotent.  `always()` removed.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:44]: It is my first time working with actions triggered by comments, so it could well be there are much better techniques.

File: .github/workflows/stage_release.yaml
Comment (SamBarker) [reply .github/workflows/stage_release.yaml:152]: We need to figure out the district list for them, as I don't think they go to everyone who is running a release

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml]: I've switch to a dedicated team (release-engineers)

File: .github/workflows/stage_release.yaml
Comment (k-wall) [inline .github/workflows/stage_release.yaml:170]: from @SamBarker 
```suggestion
                 -f "description=Blocked until staging repository is released to Maven Central." \
                 -f "context=maven-central"
```

File: .github/workflows/promote_release.yaml
Comment (k-wall) [inline .github/workflows/promote_release.yaml:171]:
```suggestion
                 -f "context=maven-central"
```

File: .github/workflows/promote_release.yaml
Comment (k-wall) [inline .github/workflows/promote_release.yaml:170]:
```suggestion
                 -f "description=Release to Maven Central is complete" \
```

File: .github/workflows/promote_release.yaml
Comment (SamBarker) [inline .github/workflows/promote_release.yaml:100]: I think the sed pattern deserves a comment (mostly because I don't remember what `^v` does off the top of my head and struggle to visualise what its ensuring. 

Though I do wonder if it should be sanitising TAG earlier to make the echo simpler... (that could well be point less shuffling)

File: .github/workflows/promote_release.yaml
Comment (SamBarker) [inline .github/workflows/promote_release.yaml:118]: Why not add the comment and auto drop the release?

File: .github/workflows/promote_release.yaml
Comment (SamBarker) [reply .github/workflows/promote_release.yaml:170]: Yes I think this change is worth applying

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [reply .github/workflows/robot-command-dispatcher.yaml:34]: Is there a reason to perform the checkout before validating the team memvership?

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:40]: Should this be `release-engineers` ?

Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:40]: Validating membership prior to extracting the command implies all commands are available to a single group. 

Can we output a list of teams for the current user and then use that to match specific groups to specific commands?

File: .github/workflows/robot-command-dispatcher.yaml
Comment (SamBarker) [inline .github/workflows/robot-command-dispatcher.yaml:86]: 😻

**Review by SamBarker — APPROVED** [review]
LGTM once the sonatype -> maven central change is merged.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:34]: I don't think so, but I don't remember clearly now.

File: .github/workflows/promote_release.yaml
Comment (k-wall) [reply .github/workflows/promote_release.yaml:100]: Version tags have a `v` prefix, so it needs that part removed.

File: .github/workflows/promote_release.yaml
Comment (k-wall) [reply .github/workflows/promote_release.yaml:118]: that would be possible, it'd just need to flip the command.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:40]: yes

Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:40]: I don't know.  I haven't experimented with the teams stuff at all yet so there may well be much smarter ways to do it.

File: .github/workflows/promote_release.yaml
Comment (k-wall) [reply .github/workflows/promote_release.yaml:79]: No, `gh` requires a checkout.

File: .github/workflows/robot-command-dispatcher.yaml
Comment (k-wall) [reply .github/workflows/robot-command-dispatcher.yaml:34]: Ok, memory has returned..  The `gh` command requires a checkout to function.

File: .github/workflows/promote_release.yaml
Comment (k-wall) [reply .github/workflows/promote_release.yaml:122]: I'll leave that as an improvement for later.

**Pull request merged**
